docs/specs/src/hwp_4_formal.md

```diff
@@ -0,0 +1,534 @@
+
+**DHT Addresses:** Both Actions and Entries from source chains can be retrieved from the DHT by either the ActionHash or EntryHash. The DHT `get()` function call returns a Record, a tuple containing the most relevant action/entry pair. Structurally, Actions "contain" their referenced entries so that pairing is obvious when a Record is retrieved by ActionHash. However, Actions are also attached as metadata at an EntryHash, and there could be many Actions which have created the same Entry content. A `get()` function called by EntryHash returns the oldest undeleted Action in the pair, while a `get_details()` function call on an EntryHash returns all of the Actions.
+
+**Agent Addresses & Agent Activity:** Technically an AgentPubKey functions as both a content address (which is never really used because performing a `get()` on the key just returns the key you used to get it) and a network address to send communications to that agents. But in addition to the content of the key stored on the DHT is metadata about that agent's chain activity. In other words, a `get_agent_activity()` request retrieves metadata about their chain records and chain status.
+
+Formally, the entire GDHT is represented as a set of 'basis hashes' $b_{c_x}$, or addresses where both content $c$ and metadata $m$ may be stored:
+
+$$
+\begin{aligned}
+GDHT &= \{d_1, \dots, d_n \} \\
+d_{b_{c_x}} &= \{c_x, M \} \\
+b_{c_x} &= hash(c_x) \\
+C &= E \bigsqcup H \\
+E &= \{e_1, \dots, e_n\} \\
+H &= \{h_1, \dots, h_n\} \\
+M &= \{m_1, \dots, m_n \} \\
+m_x &= \text{metadata}
+\end{aligned}
+$$
+
+There may be arbitrary types of metadata. Some important ones are:
+
+$$
+\begin{aligned}
+\forall e \: M_{context} &= \{{h_1}_e, \dots, {h_n}_e \} \\
+\exists c : M_{link} &= \{ link_1, \dots, link_n \} \\
+link &= \{ type, tag, b_{c_T} \} \\
+\end{aligned}
+$$
+
+For links, we refer to an address with link metadata as a **Base** and the address that the link points to as a **Target**. The link can also be typed and have an optional ${tag}$ containing arbitrary content.
+
+**Topological Transform Operations:** A source chain is a hash chain of actions with entries, but these are transformed into DHT operations which ask DHT nodes to perform certain validation and storage tasks on the content and metadata at the address, because we are transforming or projecting from authorship history to a distributed graph. Chain entries and actions are straightforwardly stored in the graph as nodes, as $C$ at their hash in the DHT, but more sophisticated operations are also performed on existing DHT entries. For example, when updating/deleting entries, or adding/removing links, additional metadata is registered in parts of the DHT to properly weave a graph.
```

Comment:
Personally I would call these Graph Transform Operations. Seems misleading to talk about the topology, because topology is a trait of graphs to do with high-level aspects of interconnectedness (presence of holes, loops, etc), whereas we really care about the actaul concrete graph.

Reply:
Ah, yeah I was focusing on the wrong thing. I *guess* that works! Though it's maybe confusing to use the same word to refer to DHT operations, and the process with produces them (the topological transform operations)